File: ts/kip-do/src/projection/policy.ts
/**
 * The Epistemic Policy a projection runs under.
 *
 * A policy is named and versioned, and every projection reports which one it
 * used (Spec §54): "accepted" with no policy attached is not an auditable
 * statement, because the next reader cannot tell what "accepted" meant.
 *
 * Overriding a threshold therefore produces a policy that is no longer the
 * baseline, and its identity changes with it. An answer reporting
 * `kip:policy:baseline` while running on different numbers would be a false
 * audit trail — worse than no trail, because it looks like one.
 */

import { errors } from '../errors.js'
import type { JsonMap } from '../json.js'

export const BASELINE_ID = 'kip:policy:baseline'

/**
 * The baseline's version.
 *
 * Bumped whenever its numbers move, because that changes what a past
 * "accepted" would have meant.
 */
export const BASELINE_VERSION = 1

export interface Policy {
  id: string
  version: number
  /** The Assertion modes this policy admits as answers. */
  modes: string[]
  /** Support at or above which a claim is accepted. */
  accept: number
  /** Support below which a side is not material to the outcome. */
  material: number
  /** What an Assertion that stated no confidence contributes. */
  unstated_confidence: number
  /** Whether support for a rival value of a functional slot opposes this one. */
  expand_conflicts: boolean
  /** How much of the Epistemic Ledger the answer carries (§49.1). */
  explanation: Explanation
}

/**
 * How much explanation a projection returns (§49.1).
 *
 * A caller that asked for `none` and got the full ledger has been handed the
 * Assertion ids, the actors and the exclusion reasons it declined — which §49.2
 * treats as a disclosure decision rather than a formatting one.
 */
export type Explanation = 'none' | 'summary' | 'ledger'

/** Every member `WITH EPISTEMIC` accepts.
 *
 * Checked exhaustively, because the alternative is what this block used to do:
 * parse `explanation: "none"`, return the whole ledger anyway, and report
 * success. A setting that is accepted and ignored is indistinguishable from one
 * that was honored.
 */
const EPISTEMIC_SETTINGS = [
  'policy',
  'accept',
  'material',
  'modes',
  'include_hypothetical',
  'include_predicted',
  'include_historical',
  'explanation',
  'purpose',
  'risk',
] as const

/**
 * Ordinary factual recall: what somebody observed, said, inferred or imported.
 *
 * `hypothetical` is entertained without commitment and `predicted` is about the
 * future, so neither is an answer to "what is the case" (§38, §39).
 */
export function baseline(): Policy {
  return {
    id: BASELINE_ID,
    version: BASELINE_VERSION,
    modes: ['observed', 'stated', 'inferred', 'imported'],
    accept: 0.7,
    material: 0.3,
    unstated_confidence: 0.5,
    expand_conflicts: true,
    explanation: 'ledger',
  }
}

/**
 * Predictions instead of observations.
 *
 * A separate policy rather than a flag, because "what is the case" and "what is
 * expected" are different questions and an answer must say which one it
 * answered.
 */
export function forecast(): Policy {
  return { ...baseline(), id: 'kip:policy:forecast', modes: ['predicted', 'inferred'] }
}

/** Reads a `WITH EPISTEMIC { … }` block. */
export function policyFromSettings(settings: JsonMap): Policy {
  const named = settings.policy
  let policy: Policy
  if (named === undefined || named === null) {
    policy = baseline()
  } else if (typeof named !== 'string') {
    throw errors.typeMismatch('`policy` must be a policy identifier')
  } else if (named === BASELINE_ID || named === 'baseline') {
    policy = baseline()
  } else if (named === 'forecast' || named === 'kip:policy:forecast') {
    policy = forecast()
  } else {
    // Naming the policy rather than defaulting to the baseline: a caller that
    // asked for a stricter reading and silently got the ordinary one would act
    // on an answer it did not request.
    throw errors.projectionPolicyUnavailable(
      `no Epistemic Policy named ${JSON.stringify(named)} is available here; ` +
        `LIST EPISTEMIC POLICIES shows the ones that are`,
    )
  }

  let overridden = false
  for (const key of ['accept', 'material'] as const) {
    const value = settings[key]
    if (value === undefined || value === null) continue
    if (typeof value !== 'number' || value < 0 || value > 1) {
      throw errors.typeMismatch(`\`${key}\` must be a number in [0, 1]`)
    }
    policy[key] = value
    overridden = true
  }
  if (policy.accept < policy.material) {
    throw errors.constraintViolation(
      'an acceptance threshold below the materiality threshold would accept ' +
        'claims it had already called immaterial',
    )
  }

  const modes = settings.modes
  if (modes !== undefined && modes !== null) {
    if (!Array.isArray(modes) || modes.some((m) => typeof m !== 'string')) {
      throw errors.typeMismatch('`modes` must be a list of Assertion modes')
    }
    policy.modes = modes as string[]
    overridden = true
  }

  // §49's own vocabulary. These name eligibility, not weighting: a
  // hypothetical admitted here is still a hypothetical, and the answer still
  // says which policy it ran under.
  for (const [key, mode] of [
    ['include_hypothetical', 'hypothetical'],
    ['include_predicted', 'predicted'],
  ] as const) {
    const value = settings[key]
    if (value === undefined || value === null) continue
    if (typeof value !== 'boolean') {
      throw errors.typeMismatch(`\`${key}\` is a boolean`)
    }
    const present = policy.modes.includes(mode)
    if (value && !present) policy.modes = [...policy.modes, mode]
    if (!value && present) policy.modes = policy.modes.filter((m) => m !== mode)
    overridden = true
  }
  const historical = settings.include_historical
  if (historical !== undefined && historical !== null) {
    if (typeof historical !== 'boolean') {
      throw errors.typeMismatch('`include_historical` is a boolean')
    }
    // Retracted and superseded Assertions are history, and history is not what
    // this Brain currently holds (§14). Admitting them into a projection would
    // let a withdrawn claim decide a present belief, which is the one thing
    // retraction is for.
    if (historical) {
      throw errors.unsupportedCapability(
        '`include_historical` would let retracted and superseded Assertions ' +
          'decide a present belief; read the history with FIND ... AS OF, ' +
          'which answers at a coordinate where those claims still stood',
      )
    }
  }
  const explanation = settings.explanation
  if (explanation !== undefined && explanation !== null) {
    if (
      typeof explanation !== 'string' ||
      !['none', 'summary', 'ledger'].includes(explanation)
    ) {
      throw errors.constraintViolation(
        '`explanation` is one of "none" | "summary" | "ledger" (§49.1), got ' +
          JSON.stringify(explanation),
      )
    }
    policy.explanation = explanation as Explanation
    overridden = true
  }
  for (const name of Object.keys(settings)) {
    if (!(EPISTEMIC_SETTINGS as readonly string[]).includes(name)) {
      throw errors.schemaFieldNotFound(
        `\`WITH EPISTEMIC\` has no setting named \`${name}\`; it takes ` +
          EPISTEMIC_SETTINGS.join(', '),
      )
    }
  }

  if (overridden) {
    policy.id = `${policy.id}+custom`
  }
  return policy
}

/** Whether a policy admits an Assertion's mode as an answer. */
export const admits = (policy: Policy, mode: string): boolean =>
  policy.modes.includes(mode)

/** Why a mode was excluded, in words a caller can act on. */
export function modeExclusion(mode: string): string {
  switch (mode) {
    case 'hypothetical':
      return 'hypothetical_not_requested'
    case 'predicted':
      return 'prediction_not_requested'
    case '':
      return 'invalid_schema'
    default:
      return 'policy_excluded'
  }
}

File: rs/anda_cognitive_nexus/src/projection/policy.rs
//! # Projection policy
//!
//! Belief is projected *under a policy*, and the policy identity travels with
//! the answer (Spec §54): "accepted" with no policy attached is not an
//! auditable statement, because the next reader cannot tell what "accepted"
//! required.
//!
//! ## What this policy deliberately does not do
//!
//! **It does not weight by mode.** A mode says how a claim was arrived at, and
//! it does not automatically grant trust (§26). Making `observed` outrank
//! `stated` here would be a trust model smuggled in as a constant — and one no
//! deployment agreed to. Modes gate *eligibility* instead: a hypothetical is
//! excluded from a factual projection because it was never offered as a fact.
//!
//! **It does not count.** Three Assertions by one actor are one voice, not
//! three (§93, §94). Aggregation runs over corroboration groups, and the policy
//! only decides where the boundaries between groups are.

use anda_kip::{AssertionMode, Json, KipError, Map};

/// The baseline policy's identity.
pub const BASELINE_ID: &str = "kip:policy:baseline";
/// The baseline policy's version. Any change to the constants below is a new
/// version, because it changes what a past "accepted" would have meant.
pub const BASELINE_VERSION: u64 = 1;

/// The knobs a projection runs under.
#[derive(Clone, Debug)]
pub struct Policy {
    /// The policy's name, reported with every answer.
    pub id: String,
    /// The policy's version.
    pub version: u64,
    /// The modes eligible for this projection.
    pub modes: Vec<AssertionMode>,
    /// The score at or above which one side is sufficient on its own.
    pub accept: f64,
    /// The score at or above which a side is *material* — enough to contest,
    /// not enough to decide.
    pub material: f64,
    /// The weight given to an Assertion whose actor stated no confidence.
    ///
    /// Not zero and not one: silence about confidence is neither a denial nor
    /// a certainty, and treating it as either would put a number in an actor's
    /// mouth (§5).
    pub unstated_confidence: f64,
    /// Whether a functional predicate's competing values oppose one another.
    pub expand_conflicts: bool,
}

impl Default for Policy {
    fn default() -> Self {
        Self::baseline()
    }
}

impl Policy {
    /// The policy a projection uses when the query names none.
    ///
    /// Ordinary factual recall: what somebody observed, said, inferred or
    /// imported. `hypothetical` is entertained without commitment and
    /// `predicted` is about the future, so neither is an answer to "what is
    /// the case" (§38, §39).
    pub fn baseline() -> Self {
        Self {
            id: BASELINE_ID.to_string(),
            version: BASELINE_VERSION,
            modes: vec![
                AssertionMode::Observed,
                AssertionMode::Stated,
                AssertionMode::Inferred,
                AssertionMode::Imported,
            ],
            accept: 0.7,
            material: 0.3,
            unstated_confidence: 0.5,
            expand_conflicts: true,
        }
    }

    /// The forecast policy: predictions instead of observations.
    ///
    /// A separate policy rather than a flag, because "what is the case" and
    /// "what is expected" are different questions and an answer must say which
    /// one it answered.
    pub fn forecast() -> Self {
        Self {
            id: "kip:policy:forecast".to_string(),
            modes: vec![AssertionMode::Predicted, AssertionMode::Inferred],
            ..Self::baseline()
        }
    }

    /// Reads a `WITH EPISTEMIC { ... }` block.
    ///
    /// Overriding a threshold produces a policy that is no longer the baseline,
    /// so the identity changes with it: an answer that reported
    /// `kip:policy:baseline` while running on different numbers would be a
    /// false audit trail.
    pub fn from_settings(settings: &Map<String, Json>) -> Result<Self, KipError> {
        let mut policy = match settings.get("policy") {
            None | Some(Json::Null) => Policy::baseline(),
            Some(Json::String(name)) => match name.as_str() {
                BASELINE_ID | "baseline" => Policy::baseline(),
                "forecast" | "kip:policy:forecast" => Policy::forecast(),
                other => {
                    return Err(KipError::new(
                        anda_kip::KipErrorCode::ProjectionPolicyUnavailable,
                        format!(
                            "this Nexus knows the epistemic policies \"baseline\" and \
                             \"forecast\"; it has no {other:?}"
                        ),
                    ));
                }
            },
            Some(other) => {
                return Err(KipError::type_mismatch(format!(
                    "`policy` names a policy, got {other}"
                )));
            }
        };

        let mut overridden = false;
        if let Some(value) = threshold(settings, "accept")? {
            policy.accept = value;
            overridden = true;
        }
        if let Some(value) = threshold(settings, "material")? {
            policy.material = value;
            overridden = true;
        }
        if let Some(modes) = settings.get("modes") {
            policy.modes = parse_modes(modes)?;
            overridden = true;
        }
        if policy.material > policy.accept {
            return Err(KipError::invalid_syntax(format!(
                "`material` ({}) must not exceed `accept` ({}); a side cannot be decisive at a \
                 lower score than it is merely material",
                policy.material, policy.accept
            )));
        }
        if overridden {
            policy.id = format!("{}+custom", policy.id);
        }
        Ok(policy)
    }

    /// Whether an Assertion's mode is eligible here.
    pub fn admits(&self, mode: Option<AssertionMode>) -> bool {
        match mode {
            // An Assertion with no recorded mode was never told how it was
            // arrived at. Admitting it would let unclassified claims decide
            // beliefs; the exclusion is reported rather than silent.
            None => false,
            Some(mode) => self.modes.contains(&mode),
        }
    }

    /// The exclusion reason a rejected mode carries (§137).
    pub fn mode_exclusion(&self, mode: Option<AssertionMode>) -> &'static str {
        match mode {
            Some(AssertionMode::Hypothetical) => "hypothetical_not_requested",
            Some(AssertionMode::Predicted) => "prediction_not_requested",
            None => "invalid_schema",
            _ => "policy_excluded",
        }
    }

    /// The policy identity reported alongside an answer.
    pub fn identity(&self) -> anda_kip::PolicyIdentity {
        anda_kip::PolicyIdentity {
            id: Some(self.id.clone()),
            version: Some(anda_kip::PolicyVersion::Integer(self.version)),
        }
    }
}

fn threshold(settings: &Map<String, Json>, key: &str) -> Result<Option<f64>, KipError> {
    match settings.get(key) {
        None | Some(Json::Null) => Ok(None),
        Some(Json::Number(n)) => {
            let value = n.as_f64().unwrap_or(f64::NAN);
            if !(0.0..=1.0).contains(&value) {
                return Err(KipError::invalid_syntax(format!(
                    "`{key}` is a score boundary in [0, 1], got {value}"
                )));
            }
            Ok(Some(value))
        }
        Some(other) => Err(KipError::type_mismatch(format!(
            "`{key}` must be a number in [0, 1], got {other}"
        ))),
    }
}

fn parse_modes(value: &Json) -> Result<Vec<AssertionMode>, KipError> {
    let Json::Array(items) = value else {
        return Err(KipError::type_mismatch("`modes` must be an array of modes"));
    };
    items
        .iter()
        .map(|item| {
            serde_json::from_value::<AssertionMode>(item.clone())
                .map_err(|_| KipError::type_mismatch(format!("{item} is not an Assertion mode")))
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn settings(value: Json) -> Map<String, Json> {
        value.as_object().cloned().unwrap()
    }

    #[test]
    fn the_baseline_excludes_what_was_never_offered_as_fact() {
        // Spec §38, §39: a hypothetical is entertained without commitment and
        // a prediction is about the future. Neither answers "what is the case".
        let policy = Policy::baseline();
        assert!(policy.admits(Some(AssertionMode::Observed)));
        assert!(policy.admits(Some(AssertionMode::Stated)));
        assert!(policy.admits(Some(AssertionMode::Imported)));
        assert!(!policy.admits(Some(AssertionMode::Hypothetical)));
        assert!(!policy.admits(Some(AssertionMode::Predicted)));
        assert_eq!(
            policy.mode_exclusion(Some(AssertionMode::Hypothetical)),
            "hypothetical_not_requested"
        );

        // Forecast asks the other question, and says so by being a different
        // policy rather than a flag on this one.
        assert!(Policy::forecast().admits(Some(AssertionMode::Predicted)));
        assert!(!Policy::forecast().admits(Some(AssertionMode::Observed)));
    }

    #[test]
    fn an_unclassified_assertion_does_not_get_to_decide() {
        assert!(!Policy::baseline().admits(None));
    }

    #[test]
    fn overriding_a_threshold_changes_the_policy_identity() {
        // Reporting `kip:policy:baseline` while running on different numbers
        // would be a false audit trail.
        let custom = Policy::from_settings(&settings(json!({"accept": 0.9}))).unwrap();
        assert_eq!(custom.accept, 0.9);
        assert_ne!(custom.id, BASELINE_ID);
        assert!(custom.id.starts_with(BASELINE_ID));

        let plain = Policy::from_settings(&settings(json!({}))).unwrap();
        assert_eq!(plain.id, BASELINE_ID);
    }

    #[test]
    fn an_incoherent_boundary_is_refused() {
        let err =
            Policy::from_settings(&settings(json!({"accept": 0.2, "material": 0.8}))).unwrap_err();
        assert!(err.message.contains("must not exceed"));
    }

    #[test]
    fn an_unknown_policy_is_named_rather_than_silently_defaulted() {
        // Falling back to the baseline would answer a question nobody asked
        // and report a policy the caller did not choose.
        let err = Policy::from_settings(&settings(json!({"policy": "strict"}))).unwrap_err();
        assert_eq!(err.name(), "ProjectionPolicyUnavailable");
    }

    #[test]
    fn silence_about_confidence_is_neither_denial_nor_certainty() {
        let policy = Policy::baseline();
        assert!(policy.unstated_confidence > 0.0 && policy.unstated_confidence < 1.0);
    }
}
